src/tlsa.rs

```diff
@@ -0,0 +1,120 @@
+//! This types in this module relate to the TLSA and DANE
+//! as described by [RFC6698](https://datatracker.ietf.org/doc/html/rfc6698)
+use crate::Der;
+
+/// Corresponds to the TLSA DNS RR Type as defined by
+/// [RFC6698](https://datatracker.ietf.org/doc/html/rfc6698)
+#[derive(Clone, Debug, PartialEq)]
+pub struct TlsaRecord<'a> {
+    /// Specifies the provided association that will be used to match
+    /// the certificate presented in the TLS handshake.
+    pub certificate_usage: CertificateUsage,
+
+    /// specifies which part of the TLS certificate presented by the
+    /// server will be matched against the association data
+    pub selector: Selector,
+
+    /// Specifies how the certificate association is presented.
+    pub matching: MatchingType,
+
+    /// This field specifies the "certificate association data" to be
+    /// matched.  These bytes are either raw data (that is, the full
+    /// certificate or its SubjectPublicKeyInfo, depending on the selector)
+    /// for matching type 0, or the hash of the raw data for matching types 1
+    /// and 2.  The data refers to the certificate in the association, not to
+    /// the TLS ASN.1 Certificate object.
+    pub association_data: Der<'a>,
+}
+
+/// A TLSA Certificate Usage as described in [RFC6698 section
+/// 7.2](https://datatracker.ietf.org/doc/html/rfc6698#section-7.2).
+/// This may be extended by future RFCs and is thus marked as
+/// non-exhaustive.
+#[repr(u8)]
+#[non_exhaustive]
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum CertificateUsage {
+    /// Certificate usage 0 is used to specify a CA certificate, or
+    /// the public key of such a certificate, that MUST be found in any of
+    /// the PKIX certification paths for the end entity certificate given
+    /// by the server in TLS.  This certificate usage is sometimes
+    /// referred to as "CA constraint" because it limits which CA can be
+    /// used to issue certificates for a given service on a host.  The
+    /// presented certificate MUST pass PKIX certification path
+    /// validation, and a CA certificate that matches the TLSA record MUST
+    /// be included as part of a valid certification path.  Because this
+    /// certificate usage allows both trust anchors and CA certificates,
+    /// the certificate might or might not have the basicConstraints
+    /// extension present
+    CAConstraint = 0,
+    /// Certificate usage 1 is used to specify an end entity
+    /// certificate, or the public key of such a certificate, that MUST be
+    /// matched with the end entity certificate given by the server in
+    /// TLS.  This certificate usage is sometimes referred to as "service
+    /// certificate constraint" because it limits which end entity
+    /// certificate can be used by a given service on a host.  The target
+    /// certificate MUST pass PKIX certification path validation and MUST
+    /// match the TLSA record.
+    ServiceCertificateConstraint = 1,
```

Comment:
To the extent that this documentation is quoting directly from the spec, perhaps make that clear by using quotes of some kind and linking to the originating section. It feels like there is a lot of boilerplate/repetition here that is making it harder to actually understand the core of the difference between these variants?